### PR TITLE
Mentor System

### LINKED ISSUE #3728: Mentors B: 30
# Mentors

## what are they gonna do? 
Mentors are the people who help other people to get familiar with game mechanics. This is done with a Mentor help chat, which is just like the Admin help chat, but is accessible by Admins _and_ mentors.

### Why do we need them? 
We need them because we lack admins and we can easily assign mentors to see if they do job good or not, they can be promoted! It's a good way to see if person can make a good ADMEME, and reduces the amount of work admins have to do.

### How will it work? 
Add mentorhelp. Same as adminhelp but as well as admins, it will be seen to people with new role - mentors. They are trusted people, familiar with the game mechanics, and willing to share this information with newbies. Only difference between normal players will be that they can answer mentorhelp and see it, and they get a (Mentor) tag on their username just like admins do.

They can be designated as mentors just like how admins are, with the server system operator copying and pasting their account ID into a mentors text file.

## what is needed

- [x] A mentor help chat that functions just like Admin help chat. Admins will need access to this as well.
- [x] The **Mentor** username tag in OOC chat
- [x] The ability to add and remove mentors from a server in a similar manner to Admins
- [ ] The discord bot webhooks must be modified to also track mentor-help logs

Your Goose. (edited by Jack to make it more concise)

### PR DESCRIPTION
**Pull request by craftxbox**
### Purpose
Implements the mentor system, as outlined in #3728
Closes #3728

### Changelog
CL: Adds mentors

### CODE REVIEW

**Comment by Bod9001** [comment]
Assets\Scripts\Managers\PlayerList.Admin.cs(244,63): error CS0117: 'Category' does not contain a definition for 'Mentor'

**Comment by craftxbox** [comment]
> Assets\Scripts\Managers\PlayerList.Admin.cs(244,63): error CS0117: 'Category' does not contain a definition for 'Mentor'

yeah I know , I haven't gotten around to anything else yet

**Comment by Bod9001** [comment]
humm, is it ready for review yet?

**Comment by craftxbox** [comment]
> humm, is it ready for review yet?

Not yet, it compiles now but all i've done is add the mentors store, nothing in the game is actually utilizing it yet

File: UnityProject/Assets/Scripts/UI/Systems/AdminTools/MentorPlayerChat.cs [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+		}
+
+		private void OnDisable()
+		{
+			chatScroll.OnInputFieldSubmit -= OnInputSend;
+		}
+
+		public void OnInputSend(string message)
+		{
+			var mentorMsg = new AdminChatMessage
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/UI/Systems/AdminTools/MentorPlayerChat.cs]: Codacy found an issue: [Remove the unused local variable 'mentorMsg'.](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6374803)

**Comment by Bod9001** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- UnityProject/Assets/Scripts/Core/Chat/Chat.cs  1
- UnityProject/Assets/Scripts/UI/Systems/AdminTools/MentorPlayerChat.cs  5
- UnityProject/Assets/Scripts/UI/Systems/AdminTools/AdminChatWindows.cs  1
- UnityProject/Assets/Scripts/Messages/Client/Admin/RequestMentorBwoink.cs  3
- UnityProject/Assets/Scripts/Messages/Server/AdminTools/MentorPlayerChatUpdateMessage.cs  1
- UnityProject/Assets/Scripts/Messages/Server/AdminTools/MentorBwoinkMessage.cs  1
- UnityProject/Assets/Scripts/UI/Systems/AdminTools/MentorHelpChat.cs  1
- UnityProject/Assets/Scripts/Messages/Server/UpdateConnectedPlayersMessage.cs  1
- UnityProject/Assets/Scripts/Messages/Client/Admin/MentorReplyMessage.cs  1
         

Clones added
============
- UnityProject/Assets/Scripts/UI/Systems/AdminTools/MentorPlayerChat.cs  4
- UnityProject/Assets/Scripts/UI/Systems/AdminTools/MentorHelpChat.cs  1
- UnityProject/Assets/Scripts/UI/Systems/AdminTools/AdminChatButtons.cs  1
- UnityProject/Assets/Scripts/Messages/Client/Admin/MentorReplyMessage.cs  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6374803&bid=20683218)

File: UnityProject/Assets/Scripts/UI/Systems/AdminTools/MentorPlayerChat.cs (hunk 1 above)
Comment (craftxbox) [reply UnityProject/Assets/Scripts/UI/Systems/AdminTools/MentorPlayerChat.cs]: shush bot this is a draft pr

**Comment by PerfectTangent** [comment]
is this still in progress?

**Comment by corp-0** [comment]
@craftxbox we haven't heard of progress on this for a while. I'm just making sure you're still working on it, otherwise just let me know so we can unassign and leave the bounty open for someone else who wants to claim it.

**Comment by craftxbox** [comment]
> @craftxbox we haven't heard of progress on this for a while. I'm just making sure you're still working on it, otherwise just let me know so we can unassign and leave the bounty open for someone else who wants to claim it.

Yeah, I'm still working on it albeit slowly, I'm still having authentication issues that are preventing me from testing and continuing work in this. Moving a bit slow because life is getting in the way atm

**Comment by craftxbox** [comment]
Doing the final tests, ready for review.

**Comment by craftxbox** [comment]
note: merge conflict label is a lie :)

**Comment by corp-0** [comment]
bot is smart enough to detect merge conflicts and label inmediately but can't get rid of the label when the conflict is solved.

File: UnityProject/Assets/Scripts/Core/Input System/KeyboardInputManager.cs [hunk 2]
```diff
@@ -220,7 +220,7 @@ private bool CheckComboEvent(KeyCombo keyCombo, KeyEventType keyEventType = KeyE
 		{ KeyAction.ChatLocal,		() => { ChatUI.Instance.OpenChatWindow(ChatChannel.Local); }},
 		{ KeyAction.ChatRadio,		() => { ChatUI.Instance.OpenChatWindow(ChatChannel.Common); }},
 		{ KeyAction.ChatOOC,		() => { ChatUI.Instance.OpenChatWindow(ChatChannel.OOC); }},
-		{ KeyAction.ToggleAHelp,    () => { ChatUI.Instance.OnAdminHelpButton(); }},
+		{ KeyAction.ToggleAHelp,    () => { ChatUI.Instance.OnHelpButton(); }},
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Core/Input System/KeyboardInputManager.cs:223]: what window does this show?

Comment (craftxbox) [reply UnityProject/Assets/Scripts/Core/Input System/KeyboardInputManager.cs:223]: It's showing a window with two buttons  one for ahelp and another for mhelp, let me grab a picture

Comment (craftxbox) [reply UnityProject/Assets/Scripts/Core/Input System/KeyboardInputManager.cs:223]: https://crxb.cc/gxva07d top right of here is what's shown

**Comment by craftxbox** [comment]
just to note  while I have already implemented them in code, I haven't tested the mentor ooc tags , nor have I tested mentor chat to discord webhook and won't be able to check them for about 15 hours

**Comment by corp-0** [comment]
most of this is a copy of how the admin chat does it, isn't it? It *should* work but still, I will wait until someone who knows the networking better to give some input.

**Comment by craftxbox** [comment]
practically, yes, there didn't seem like a better way to do it short of rewriting the whole thing 

well, I have tested the main functionality to ensure it works, and atleast on my machine', it does

**Comment by craftxbox** [comment]
still ready for review but DNM, encountered a problem and unity is, again, telling me to fuck myself. ;-;

**Comment by craftxbox** [comment]
mergable :)

**Comment by corp-0** [comment]
merging

**Pull request merged**